pkg/core/state/contract.go

```diff
@@ -93,12 +93,16 @@ func (c *Contract) FromStackItem(item stackitem.Item) error {
 	if err != nil {
 		return err
 	}
-	c.NEF, err = nef.FileFromBytes(bytes)
+	skipCheck := false
+	if c.UpdateCounter > 0 {
+		skipCheck = true
+	}
+	c.NEF, err = nef.FileFromBytes(bytes, skipCheck)
 	if err != nil {
 		return err
 	}
 	m := new(manifest.Manifest)
-	err = m.FromStackItem(arr[4])
+	err = m.FromStackItem(arr[4], skipCheck)
```

Comment:
We actually have contracts cache that is responsible for contracts retrieval in native Management. So it's not a bottleneck for us. The only place where the real stored contract deserialization is being performed is contracts cache initialization:
https://github.com/nspcc-dev/neo-go/blob/a81dc5c7957093062ea13de24c6c50ede1926d92/pkg/core/native/management.go#L615-L631

But even during initialization we're 100% sure that contracts have proper state and no state diff could happen due to the fact that this check is missing in C# node.

@roman-khimov, do you allow to keep the things as is (not to remove NEF deserialization checks)? It won't affect state, and I think we can afford to spend some time during Management cache initialization. Otherwise we need to change the existing marshallers, it's just not worth the efforts.

Reply:
https://github.com/nspcc-dev/neo-go/issues/3346#issuecomment-1986836335

Even if we limit the cache (we're likely to eventually) it's not be something for us to do often. Doing this for every call (like in C#) is of course resource-consuming. With almost any kind of cache it's just not a problem.